python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py

```diff
@@ -241,3 +241,31 @@ def test_update_fixed_width_to_dynamic_strings(lib_type, request):
     expected_df = dynamic_strings_data
     read_df = lib.read(symbol).data
     assert_frame_equal(expected_df, read_df)
+
+
+# https://github.com/man-group/ArcticDB/issues/767
+# Batch write and append call aggregator_set_data from threads running in the CPU threadpool (i.e. not the main thread)
+# With unicode strings, a PyObject allocation is needed, and therefore the thread must be holding the GIL
+# This test ensures that the correct thread is holding the GIL when performing these allocations, and that there is no
+# deadlock.
+# This is not an issue with non-batch methods as aggregator_set_data is called from the main thread, which pybind11
```

Comment:
I'm not convinced this is the case, non-batch writes do the slicing in the main thread, but then write in parallel to the individual segments, calling aggregator_set_data. Is it worth copying and pasting this test into a non-batch version just to check that it doesn't deadlock?

Reply:
I might be missing something, but this seems to be before the parallelisation begins?
https://github.com/man-group/ArcticDB/blob/master/cpp/arcticdb/pipeline/write_frame.cpp#L96

I did test that it didn't deadlock with `lmdb_version_store_tiny_segment` and unicode strings with both `write` and `append`. With the debugger I could see the `scoped_gil_lock` successfully acquiring the GIL, which must have meant it was already held by the calling thread?

Reply:
Yeah you're correct, I had forgotten exactly where the parallelization started. That's not ideal actually, if there is work to be done in aggregator_set_data for other object types we might need to fix it, by moving the work into the parallelized bit. Probably some write-time equivalent of the read handlers would be useful both in achieving that and finishing the job of moving the Python-specific bits out of the core.

Reply:
Yeh, there's some non-trivial work to do in that loop for populating the string pools at least, so parallelisation could bring some benefit. Should be pretty trivial as well. I'll add a ticket.